### ISSUE TITLE
An implementation feature like torchtext.data.BucketIterator in regular Dataloader 

### ISSUE THREAD

**Issue by maroxtn** [pytorch/text#1253]
## 🚀 Feature
A way to pool sequences of similar lengths together in Dataloader, same as torchtext.data.BucketIterator.
**Motivation**

The BucketIterator feature is very helpful in torchtext data loading, it saves a lot of time during training. However in Dataloader this feature doesn't exist, it would be very practical. 

Also the torchtext API is mainly design for NLP tasks, meanwhile this feature would be useful to any sequential data.

**Comment by zhangguanheng66** [comment]
We include a code snippet for the legacy BucketIterator in our legacy migration tutorial ([link](https://github.com/pytorch/text/blob/master/examples/legacy_tutorial/migration_tutorial.ipynb)). Of course, this could be re-implemented in our library.

**Comment by maroxtn** [comment]
Thanks for the response. I also managed to implement it in a similar fashion but using a custom Sampler:

    class KSampler(Sampler):

        def __init__(self, data_source, batch_size):
            self.lens = [x[1] for x in data_source]
            self.batch_size = batch_size

        def __iter__(self):

            idx = list(range(len(self.lens)))
            arr = list(zip(self.lens, idx))

            random.shuffle(arr)
            n = self.batch_size*100

            iterator = []

            for i in range(0, len(self.lens), n):
                dt = arr[i:i+n]
                dt = sorted(dt, key=lambda x: x[0])

                for j in range(0, len(dt), self.batch_size):
                    indices = list(map(lambda x: x[1], dt[j:j+self.batch_size]))
                    iterator.append(indices)

            random.shuffle(iterator)
            return iter([item for sublist in iterator for item in sublist])  #Flatten nested list

        def __len__(self):
            return len(self.lens)

**Comment by morgankohler** [comment]
How are you declaring your DataLoader with this?

**Comment by cpuhrsch** [comment]
cc @parmeet

**Comment by parmeet** [comment]
The datasets in torchtext are iterators. Unfortunately, the idea similar to Bucket Iterator cannot be implemented on top of iterators as we can only access elements sequentially. We need to materialize the data and we have to do it efficiently for large datasets. I guess we can potentially provide functionality similar to bucket iterator as  `functional` once we figure out efficient way to materialize the data. We have an ongoing issue related to this (https://github.com/pytorch/text/issues/1296)
